module/icp/algs/sha2/sha2.c

```diff
@@ -53,17 +55,220 @@ static void Encode(uint8_t *, uint32_t *, size_t);
 static void Encode64(uint8_t *, uint64_t *, size_t);
 
 /* userspace only supports the generic version */
-#if	defined(__amd64) && defined(_KERNEL)
-#define	SHA512Transform(ctx, in) SHA512TransformBlocks((ctx), (in), 1)
-#define	SHA256Transform(ctx, in) SHA256TransformBlocks((ctx), (in), 1)
+#if	defined(_KERNEL)
+
+#include <sys/simd.h>
+
+typedef void (*sha256_block_f)(uint32_t *state, const void *in, size_t num);
+typedef void (*sha512_block_f)(uint64_t *state, const void *in, size_t num);
+
+#endif	/* _KERNEL */
+
+#if defined(_KERNEL)
+
+#define	BENCHMARK_BUFFER_SIZE (32 * 1024)
+
+// SHA-256
+
+static void sha256_alg_impl_benchmark(
+    const alg_impl_ops_t *ops, void *buffer, size_t buffer_n)
+{
+	SHA2_CTX sha2_ctx;
+	sha256_block_f sha256_impl = ops->ctx;
+	sha256_impl(sha2_ctx.state.s32, buffer, buffer_n >> 6);
+}
+
+static void
+SHA256TransformBlocksGeneric(uint32_t *state, const void *in, size_t num);
+
+static alg_impl_ops_t sha256_impl_generic = {
```

Comment:
All of these can be `static const alg_impl_ops_t *_impl_*` AFAICT

Reply:
Correct.

Fixed.